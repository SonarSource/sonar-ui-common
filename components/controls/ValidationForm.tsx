/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Formik, FormikActions, FormikProps } from 'formik';
import * as React from 'react';

export type ChildrenProps<V> = Omit<FormikProps<V>, 'handleSubmit'>;

interface Props<V> {
  children: (props: ChildrenProps<V>) => React.ReactNode;
  initialValues: V;
  isInitialValid?: boolean;
  onSubmit: (data: V) => Promise<void>;
  validate: (data: V) => { [P in keyof V]?: string } | Promise<{ [P in keyof V]?: string }>;
}

export default class ValidationForm<V> extends React.Component<Props<V>> {
  mounted = false;

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleSubmit = (data: V, { setSubmitting }: FormikActions<V>) => {
    const result = this.props.onSubmit(data);
    const stopSubmitting = () => {
      if (this.mounted) {
        setSubmitting(false);
      }
    };

    if (result) {
      result.then(stopSubmitting, stopSubmitting);
    } else {
      stopSubmitting();
    }
  };

  render() {
    return (
      <Formik<V>
        initialValues={this.props.initialValues}
        isInitialValid={this.props.isInitialValid}
        onSubmit={this.handleSubmit}
        validate={this.props.validate}>
        {({ handleSubmit, ...props }) => (
          <form onSubmit={handleSubmit}>{this.props.children(props)}</form>
        )}
      </Formik>
    );
  }
}
