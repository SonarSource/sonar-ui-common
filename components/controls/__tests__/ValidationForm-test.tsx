/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import ValidationForm from '../ValidationForm';

it('should render and submit', async () => {
  const render = jest.fn();
  const onSubmit = jest.fn();
  const setSubmitting = jest.fn();
  const wrapper = shallow(
    <ValidationForm initialValues={{ foo: 'bar' }} onSubmit={onSubmit} validate={jest.fn()}>
      {render}
    </ValidationForm>
  );
  expect(wrapper).toMatchSnapshot();
  wrapper.dive();
  expect(render).toBeCalledWith(
    expect.objectContaining({ dirty: false, errors: {}, values: { foo: 'bar' } })
  );

  wrapper.prop<Function>('onSubmit')({ foo: 'bar' }, { setSubmitting });
  expect(setSubmitting).toBeCalledWith(false);

  onSubmit.mockResolvedValue(undefined).mockClear();
  setSubmitting.mockClear();
  wrapper.prop<Function>('onSubmit')({ foo: 'bar' }, { setSubmitting });
  await new Promise(setImmediate);
  expect(setSubmitting).toBeCalledWith(false);
});
