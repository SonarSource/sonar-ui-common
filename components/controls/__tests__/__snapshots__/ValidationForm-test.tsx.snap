// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render and submit 1`] = `
<Formik
  enableReinitialize={false}
  initialValues={
    Object {
      "foo": "bar",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  validate={[MockFunction]}
  validateOnBlur={true}
  validateOnChange={true}
>
  <Component />
</Formik>
`;
